bpf/tests/tunnel_vxlan_helpers_tests.c

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#include "common.h"
+#include "bpf/ctx/skb.h"
+#include "pktgen.h"
+
+#define TUNNEL_PROTOCOL TUNNEL_PROTOCOL_VXLAN
+#define TUNNEL_PORT 8472
+#define TUNNEL_PORT_BAD 0
+#define VXLAN_VNI 0xDEADBE
+#define VXLAN_VNI_NEW 0xCAFEBE
+#define UDP_CHECK 0xDEAD
+
+#include "node_config.h"
+#include "lib/common.h"
+
+static __always_inline void
+mk_data(const __u8 *buff) {
+	struct ethhdr *eth = (struct ethhdr *)buff;
+
+	memcpy(&eth->h_source, (__u8 *)mac_one, sizeof(mac_three));
+	memcpy(&eth->h_dest, (__u8 *)mac_one, sizeof(mac_four));
+	eth->h_proto = ETH_P_IP;
+
+	struct iphdr *ipv4 = (struct iphdr *)(buff + sizeof(struct ethhdr));
+
+	ipv4->saddr = v4_pod_one;
+	ipv4->daddr = v4_pod_two;
+}
+
+static __always_inline int
+mk_packet(struct __ctx_buff *ctx) {
+	struct pktgen builder;
+	struct udphdr *l4;
+	struct vxlanhdr *vx;
+	/* data is encap'd ipv4 packet, we don't care about l4 */
+	__u8 encap_data[sizeof(struct ethhdr) + sizeof(struct iphdr)];
+	void *data;
+
+	pktgen__init(&builder, ctx);
+
+	l4 = pktgen__push_ipv4_udp_packet(&builder,
+					  (__u8 *)mac_one,
+					  (__u8 *)mac_two,
+					  v4_node_one,
+					  v4_node_two,
+					  666,
+					  bpf_htons(TUNNEL_PORT));
+	if (!l4)
+		return TEST_ERROR;
+
+	l4->check = UDP_CHECK;
+
+	vx = pktgen__push_default_vxlanhdr(&builder);
+	if (!vx)
+		return TEST_ERROR;
+
+	vx->vx_vni = bpf_htonl(VXLAN_VNI << 8);
+
+	mk_data(encap_data);
```

Comment:
How does this differ from `pktgen__push_ipv4_packet()` ?

Reply:
I tried this at first, but due to packet layer accounting, it did not work.

Reply:
Could you add a comment to that effect, so we know to clean it up when the layer-count gets bumped?

Reply:
Fixed in latest push